src/themis/secure_cell_seal_passphrase.h

```diff
@@ -0,0 +1,291 @@
+ * |          iteration count          |   salt length   |                 |
+ * +--------+--------+--------+--------+--------+--------+ - - - -+ - - - -+
+ * |                       salt goes in every field                        |
+ * + - - - -+ - - - -+ - - - -+ - - - -+ - - - -+ - - - -+--------+--------+
+ * |                                                     |
+ * +--------+--------+--------+--------+--------+--------+
+ * ```
+ *
+ * @see Themis RFC 2 contains detailed reference on data formats.
+ */
+
+/**
+ * Authentication token used by Seal and Token Protect modes with passphrases.
+ *
+ * This struct provides only API.
+ * It does **not** represent actual in-memory data layout.
+ */
+struct themis_scell_auth_token_passphrase {
+    uint32_t alg;
+    const uint8_t* iv;
+    uint32_t iv_length;
+    const uint8_t* auth_tag;
+    uint32_t auth_tag_length;
+    uint32_t message_length;
+    const uint8_t* kdf_context;
+    uint32_t kdf_context_length;
+};
+
+static const size_t themis_scell_auth_token_passphrase_min_size = 5 * sizeof(uint32_t);
+
+static inline size_t themis_scell_auth_token_passphrase_size(
+    const struct themis_scell_auth_token_passphrase* hdr)
+{
+    size_t total_size = 0;
+    total_size += sizeof(hdr->alg);
+    total_size += sizeof(hdr->iv_length) + hdr->iv_length;
+    total_size += sizeof(hdr->auth_tag_length) + hdr->auth_tag_length;
+    total_size += sizeof(hdr->message_length);
+    total_size += sizeof(hdr->kdf_context_length) + hdr->kdf_context_length;
+    return total_size;
+}
+
+/*
+ * KDF context is not written to the buffer.
+ * Call KDF-specific function to write the context into the buffer.
+ */
+static inline themis_status_t themis_write_scell_auth_token_passphrase(
+    const struct themis_scell_auth_token_passphrase* hdr, uint8_t* buffer, size_t buffer_length)
+{
+    if (buffer_length < themis_scell_auth_token_passphrase_size(hdr)) {
+        return THEMIS_BUFFER_TOO_SMALL;
+    }
+    buffer = write_uint32LE(buffer, hdr->alg);
+    buffer = write_uint32LE(buffer, hdr->iv_length);
+    buffer = write_uint32LE(buffer, hdr->auth_tag_length);
+    buffer = write_uint32LE(buffer, hdr->message_length);
+    buffer = write_uint32LE(buffer, hdr->kdf_context_length);
+    buffer = write_bytes(buffer, hdr->iv, hdr->iv_length);
+    buffer = write_bytes(buffer, hdr->auth_tag, hdr->auth_tag_length);
+    /* KDF context is written separately */
+    return THEMIS_SUCCESS;
+}
+
+/*
+ * KDF context buffer is retrieved into `hdr.kdf_context`, but you need to
+ * parse it separately based on the KDF function used (see `alg` field).
+ */
+static inline themis_status_t themis_read_scell_auth_token_passphrase(
+    const uint8_t* buffer, size_t buffer_length, struct themis_scell_auth_token_passphrase* hdr)
+{
+    size_t need_length = themis_scell_auth_token_passphrase_min_size;
+    if (buffer_length < need_length) {
+        return THEMIS_FAIL;
+    }
+    hdr->alg = read_uint32LE(&buffer);
```

Comment:
first time when I read this block of code, I was confused by read 5 times of the same piece of memory. Then I pressed ctrl+f "read_uint32LE" to check that there changed pointer to buffer. Looks a little bit unusual that read operation make changes in input buffer. It expected when work with streams or files. I tried to google some usual function names for same logic, but found only api to work with files/stream, like fread, fscanf. What about to rename function like `stream_read_uint32LE` | `sread_uint32LE` | `read_next_uint32LE`? Somehow mark in function name that pointer in buffer will be changed. What do you think about it?

Reply:
Yeah, it's a good idea to add some `stream` to these function names. Obviously, I wasn't confused by this naming, but evidently first-time readers can be confused. Thank you for your valuable feedback.

The idea for these function comes from FreeRDP with which I've been working for some time. They use [`wStream`](https://github.com/FreeRDP/FreeRDP/blob/89913624beaac0da4d8e4a416509efef4a388aea/channels/rdpsnd/client/rdpsnd_main.c#L236-L245) for parsing, with function names like `Stream_Read_UINT32()`.

For some reason I thought it'd be okay to drop the 'stream' prefix, maybe because we pass a pointer and length separately or something. Now I think I was wrong and should have kept the prefix.

Reply:
I tried to find some examples where it used in such way to look how it named (to not solve second hardest problem in programming) and didn't found *( 
now will know about freeRDP ))

Reply:
I like the idea of adding `stream` prefix/suffix/whatever